### PR TITLE
Add particle system overlay in edit.js

### PR DESCRIPTION
**Pull request by huffman**
This PR adds particle system icons so you can see and click on particle systems while using edit.js. This works in the same way as lights currently do. Showing light and particle systems is now on by default.

This PR also fixes a bug where switching between point and spot lights would not update the icon.

The icons looks like:
light_particle_icons

### CODE REVIEW

**Comment by huffman** [comment]
To test:

 1. Go into edit mode
 1. Create lights and particle systems
 1. Confirm that you can see icons on top of those new entities
 1. Confirm that you can click on an icon to select that entity
 1. Confirm that you can move the entity around and the icon follows
 1. Confirm that if you change a light from a spot light to a point light (or vice versa) the icon updates
 1. Confirm that if you delete an entity the icon disappears
 1. Confirm that the icons are only shown while in edit mode

**Comment by hifi-gustavo** [comment]
Build console output for commit: c41ae0ccfae11f2e2c68856434acd32a971d7f17 available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-c41ae0c.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-c41ae0c.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: ccf16ebc882e8aa497b0da32c6ee9a9aff4b7e44 available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-ccf16eb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-ccf16eb.dmg

**Comment by nbq** [comment]
Testing this

**Comment by nbq** [comment]
Tester: Coal
System: i7 6700k, GeForce 970, Windows 10

* Did clean install
* Went into edit mode
* Created a light and a particle entity
* Checked for icons above the new entities and they are not there
* ^ According to what I can see from this PR, this one is a fail unless I am doing something wrong - See screen shot below

hifi-snap-by--on-2017-02-21_23-23-31

**FAIL**

**Comment by huffman** [comment]
@nbq Hmm, do you have `Show Lights and Particle Systems in Edit Mode` under the `Edit` menu? It should be there, and be selected.

**Comment by nbq** [comment]
@huffman yes I tried even turning it off and on and it still did not show for me

**Comment by huffman** [comment]
@nbq Ok, I figured it out, there were some missing images. Can you try again once the next build pops up?

**Comment by hifi-gustavo** [comment]
Build console output for commit: ee8a51a11d3300df1af2a316012f728a1f13fe8d available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-ee8a51a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-ee8a51a.dmg

**Comment by nbq** [comment]
retesting

**Comment by nbq** [comment]
Tester: Coal
System: i7 6700k, GeForce 970, Windows 10

* Went into edit mode
* Created a light and a particle entity
* Checked for icons above the new entities - they are there now!
* Confirmed I can click on the icon to select the entity - Fail

## Explanation of how it fails:

* Create particle entity, it is selected by default
* Select something else
* Try to click particle entity via the icon and it fails to select it
* Next close edit all together
* Open Edit again and try to click on the particle icon - it works then and then alone
* Next try to select something else
* Now try to select the particles again via the icon, it fails to select it

TLDR: It only selects the particles / light entity if its the first thing you select after entering edit mode.

**FAIL**

**Comment by huffman** [comment]
@nbq thanks for testing again. I'll try this - when I tested I clicked a bunch of different entities and didn't see any issues, but I may have been primarily clicking between light and particle entities.

**Comment by huffman** [comment]
@nbq Does this happen when the light or particle entity is behind another entity? I didn't have any issues unless the actual entity was behind another, even though you could see the icon. This is something I can look into addressing, I just want to make sure that is the issue.

**Comment by nbq** [comment]
@huffman I am hating how long it takes github to send me notifications!

To address your question, am standing in the "home" location of the sandbox and create a particle right in front of me. I have deleted the particles and lights manually (from previous tests) and tried a particle only so nothing should be in front of it. I am testing this in desktop mode using 3rd person view so I am looking over my shoulder per se. 

I know the icon is clickable since it works if I go into edit mode and it is the first thing I select. If something was in front of it then I do not think I would be able to click it after entering edit mode for the first time.

**Comment by huffman** [comment]
@nbq Ok, I think I may have replicated what you are seeing.

Here I have a light selected:
icon_fail_1

Here I have selected the main garden model:
icon_fail_2

The light's icon is now inside of the red wireframe cube showing the selected garden model, so when you click now you are hitting the wireframe overlay instead of the icon.

Does this sound right to you? Is this what you are seeing?

**Comment by nbq** [comment]
@huffman I think this is exactly what I am seeing, yes

File: scripts/system/libraries/entityIconOverlayManager.js [hunk 1]
```diff
@@ -79,24 +83,32 @@ LightOverlayManager = function() {
...
             entityIDs[entityID] = entityID;
-            Overlays.editOverlay(overlay, {
+            var overlayProperties = {
                 position: properties.position,
-                url: properties.isSpotlight ? SPOT_LIGHT_URL : POINT_LIGHT_URL,
                 rotation: Quat.fromPitchYawRollDegrees(0, 0, 270),
                 visible: visible,
                 alpha: 0.9,
                 scale: 0.5,
+                drawInFront: true,
```
Comment (Atlante45) [inline scripts/system/libraries/entityIconOverlayManager.js]: I'm not sure about this.
If there's a light/particles behind a wall, I don't need/want to see the icon.
I won't be able to select it anyway.

Comment (huffman) [inline scripts/system/libraries/entityIconOverlayManager.js]: Couple thoughts:

 1. There is a precedent for doing this, although in other editors that I've seen do this the areas that are occluded are grayed out, so this is a bit of a poor-man's version.
 2. I think there are benefits to always seeing them, in particular because lights+particles have effects that can be in my view without the position of the entity being in the front of my view. For instance, a particle system inside an object can emit particles that I can see, but the position of the particle system itself will be occluded by the object that it is in.
 3. I agree the selection is an issue. IMO we should be preferring drawInFront items when doing ray picks, which would be a solution to Jason's issue.

In any case, this is mostly subjective. I can change this back, which will fix the issue Jason is seeing.

Comment (Atlante45) [inline scripts/system/libraries/entityIconOverlayManager.js]: How far around you do we scan for entities?

Comment (huffman) [inline scripts/system/libraries/entityIconOverlayManager.js]: For these icon overlays they will show up for all known lights + particle effects in the entity tree. For the entity list it is, I think, 100 meters by default. That can be changed though.

Not sure if that answers your question.

Comment (Atlante45) [inline scripts/system/libraries/entityIconOverlayManager.js]: Oh right, we use the value entered on the UI.
I see your point, sometimes, the center of an entity will not be in view but its effect will.
Those should be contained within the entity's bounding box though.
Ideally, we would do float on top but only in the case where the bounding box is within view.
But I'm pretty sure we can't do that right now.

One of the issues I'm foreseeing with this is that I'll see a light icon, click on it, but I'll never be able to select it because it's fully behind the wall.

Comment (huffman) [inline scripts/system/libraries/entityIconOverlayManager.js]: Update:

`drawInFront: false` is not rendering images correctly. I've opened a bug report for it. In the meantime we will have to go with `drawInFront: true` and fix the issue with not being able to click a light/particle icon if there is something in front of it, despite it being visible.

Once the drawInFront issue is fixed, we can revisit the design decision around whether the icons should be drawn in front or not.

**Comment by RyanDowne** [comment]
When I test this, and select the particle by clicking on the overlay, I get the bounding box but I don't get the tab with the particle props.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8a1bd781f083a42dab34ba93fb713d0f041e3872 available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-8a1bd78.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-8a1bd78.dmg

**Comment by huffman** [comment]
@RyanDowne Thanks, I've fixed that now.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8246dbafabe8e529688c572ba253af0033211f44 available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-8246dba.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-8246dba.dmg

**Comment by Nex-Pro** [comment]
#9172


Add particle system overlay in edit.js

Tested on Win10 desktop/HMD.


Desktop mode: ### PASS

HMD mode: ### FAIL


TESTING routine:

Desktop mode: **PASS**

Confirm that you can see icons on top of those new entities:
    Particle icon visible: **Success**
    Light icon visible: **Success**

Confirm that you can click on an icon to select that entity:
    Particle icon click to select: **Success**
    Light icon click to select: **Success**

Confirm that you can move the entity around and the icon follows:
    Particle icon follows: **Success**
    Light icon follows: **Success**

Confirm that if you change a light from a spot light to a point light (or vice versa) the icon updates:
    Light Icons update between types as appropriate: **Success**
    Confirm that if you delete an entity the icon disappears:
    Delete particle, icon disappears: **Success**
    Delet light, icon disappears: **Success**

Confirm that the icons are only shown while in edit mode:
    Particle icon only shows in edit mode: **Success**
    Light Icon ( ooint ) only shows in edit mode: **Success**
    Light Icon ( spot ) only shows in edit mode: **Success**

Confirm that you can turn the icons on and off by toggling them in the Edit menu (   Show Lights and Particle Systems in Edit Mode): **Success**


Confirm that the particle explorer comes up when you select a particle system:
    Particle system tab shows when particle selected: **Success**
    Particle system tab hides when no particle is selected: **Success**


#### HMD:
    Unable to select icons in edit mode: **FAIL**

    As selecting these ents via Icon is key I did not test HMD mode further, 
    apart from icons show/don't show via menu. : **Success**


Observations:

    Particle are almost invisible against equally dark backgrounds. Needs two tone icon.

    Might make sense to be able to drag the lights or particles using the icon.

**Comment by huffman** [comment]
@Nex-Pro I added the ability to select the light and particle icons using the hand controllers. Please give this another go! Thanks!

**Comment by hifi-gustavo** [comment]
Build console output for commit: df644edde4f2048276201bc648f1438e67b85547 available at: https://gist.github.com/e1f69843ec9019994fb2568b70bce216<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-df644ed.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9712-df644ed.dmg

**Comment by Nex-Pro** [comment]
Cheking this now @huffman

**Comment by Nex-Pro** [comment]
Tested on Win10 pro
I7
Gtx 1070
HTC Vive.


### HMD:


**Looks like a PASS but did crash (may not be due to this pr)**

**See notes**

Go into edit mode: **DONE**

Create lights and particle systems: **DONE**

Confirm that you can see icons on top of those new entities: **SUCCESS**

Confirm that you can click on an icon to select that entity: **SUCCESS**

Confirm that you can move the entity around and the icon follows: **SUCCESS**

Confirm that if you change a light from a spot light to a point light (or vice versa) the icon updates: **SUCCESS**

Confirm that if you delete an entity the icon disappears: **SUCCESS**

Confirm that the icons are only shown while in edit mode: **SUCCESS**

Confirm that you can turn the icons on and off by toggling them in the Edit menu (Show Lights and Particle Systems in Edit Mode): **SUCCESS**

Confirm that the particle explorer comes up when you select a particle system: **SUCCESS**


### Desktop:

Go into edit mode: **DONE**

Create lights and particle systems: **DONE**

Confirm that you can see icons on top of those new entities: **SUCCESS**

Confirm that you can click on an icon to select that entity: **SUCCESS**

Confirm that you can move the entity around and the icon follows: **SUCCESS**

Confirm that if you change a light from a spot light to a point 
light (or vice versa) the icon updates: **SUCCESS**

Confirm that if you delete an entity the icon disappears: **SUCCESS**

Confirm that the icons are only shown while in edit mode: **SUCCESS**

Confirm that you can turn the icons on and off by toggling them in the Edit menu (Show Lights and Particle Systems in Edit Mode): **SUCCESS**

Confirm that the particle explorer comes up when you select a particle system: **SUCCESS**


### Notes: 

When Particle and light entity icons are in close proximity it can be difficult to discern which has been selected. Therefore it be a helpful addition to have two-state icons. Thus the user would be informed as to which icon is currently selected.

It would make sense to be able to drag the particles and lights around by the icon. This seems more natural than having to then try to select some random point inside the bounding box.

## SMOKE TEST

## QA Test
**QA Tester**: Piper.Peppercorn
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** SUCCESS
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**dev-playa 'playaPerformanceTest.js' Result**: **dev-playa 'playaPerformanceTest.js'** SUCCESS
**dev-demo 'playaPerformanceTest.js' Result**: **dev-demo 'playaPerformanceTest.js'** N/A
**QA Pass**: **QA Pass** SUCCESS

**Additional Test Results + Observations**:

**Domain** |localhost |dev-playa |
---|---|---|---
**Render Rate**: |89 |89 |
**Present Rate**: |90 |90 |
**Present New Rate**: |89 |89 |
**CPU Buffer Count**: |1883 |3326 |
**GPU Buffer Count**: |658 |829 |
**CPU Memory Usage**: |21906960 |56397328 |
**GPU Memory Usage**: |12306544 |30795888 |
**CPU Texture Count**: |228 |241 |
**GPU Texture Count**: |179 |170 |
**GPU Transfer Count**: |0 |0 |
**CPU Texture Memory Usage**: |106141172 |232975852 |
**GPU Texture Memory Usage**: |980227694 |975145704 |
**GPU Virtual Memory Usage**: |980227694|975145704|
**Triangle Count**: |97478 |1507707 |
**Triangle Rate**: |8861636 |150770700 |
**API Drawcall Count**: |47 |313 |
**Drawcall Count**: |80 |896 |
**Drawcall Rate**: |7272 |89600 |
**State Changes Texture Count**:|76 |298 |
**Pipeline Count**: |30 |40 |
**Item Opaques**: |34 |484 |
**Item Translucents**: |12 |25 |
**Item Lights**: |2 |41 |
**Timing Opaques**: |0.167 |1.176 |
**Timing Translucents**: |0.041 |0.046 |
**Timing Lighting**: |0.096 |0.018 |
**Render Frame**: |1.113 |2.288 |


### Crash:

Crashed at welcome while preparing results. May not be specific to this PR.

**Just in time debugger VS2013**

**Call Stack**

>	ntdll.dll!00007ff83f508283()	Unknown
 	ntdll.dll!00007ff83f508bda()	Unknown
 	ntdll.dll!00007ff83f4b5b9a()	Unknown
 	ntdll.dll!00007ff83f44c895()	Unknown
 	msvcr120.dll!00007ff82d1469d8()	Unknown
 	Qt5Core.dll!000000005e16905a()	Unknown
 	Qt5Core.dll!000000005e1765d6()	Unknown
 	interface.exe!Font::splitLines(const QString & str) Line 111	C++
 	interface.exe!Font::tokenizeForWrapping(const QString & str) Line 116	C++
 	interface.exe!Font::rebuildVertices(float x, float y, const QString & str, const glm::tvec2<float,0> & bounds) Line 274	C++
 	interface.exe!Font::drawString(gpu::Batch & batch, float x, float y, const QString & str, const glm::tvec4<float,0> * color, EffectType effectType, const glm::tvec2<float,0> & bounds, bool layered) Line 358	C++
 	interface.exe!TextRenderer3D::draw(gpu::Batch & batch, float x, float y, const QString & str, const glm::tvec4<float,0> & color, const glm::tvec2<float,0> & bounds, bool layered) Line 79	C++
 	interface.exe!RenderableTextEntityItem::render(RenderArgs * args) Line 87	C++
 	interface.exe!render::renderStateSortShapes(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext, const std::shared_ptr<render::ShapePlumber> & shapeContext, const std::vector<render::ItemBound,std::allocator<render::ItemBound> > & inItems, int maxDrawnItems, const render::ShapeKey & globalKey) Line 121	C++
 	interface.exe!DrawStateSortDeferred::run::__l11::<lambda>(gpu::Batch & batch) Line 318	C++
 	interface.exe!gpu::doInBatch<void <lambda>(gpu::Batch &) >(std::shared_ptr<gpu::Context> context, DrawStateSortDeferred::run::__l11::void <lambda>(gpu::Batch &) f) Line 296	C++
 	interface.exe!DrawStateSortDeferred::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext, const render::VaryingSet2<std::vector<render::ItemBound,std::allocator<render::ItemBound> >,std::shared_ptr<LightingModel> > & inputs) Line 325	C++
 	interface.exe!render::Job::Model<DrawStateSortDeferred,DrawStateSortConfig,render::VaryingSet2<std::vector<render::ItemBound,std::allocator<render::ItemBound> >,std::shared_ptr<LightingModel> >,render::JobNoIO>::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext) Line 558	C++
 	interface.exe!render::Job::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext) Line 585	C++
 	interface.exe!render::Task::Model<RenderDeferredTask,render::TaskConfig>::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext) Line 627	C++
 	interface.exe!render::Job::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext) Line 585	C++
 	interface.exe!render::Task::run(const std::shared_ptr<render::SceneContext> & sceneContext, const std::shared_ptr<render::RenderContext> & renderContext) Line 697	C++
 	interface.exe!Application::displaySide(RenderArgs * renderArgs, Camera & theCamera, bool selfAvatarOnly) Line 5121	C++
 	interface.exe!Application::paintGL() Line 2296	C++
 	interface.exe!Application::event(QEvent * event) Line 2514	C++
 	Qt5Widgets.dll!000000005ea47d12()	Unknown
 	Qt5Widgets.dll!000000005ea46c2f()	Unknown
 	Qt5Core.dll!000000005e26f689()	Unknown
 	Qt5Core.dll!000000005e27140e()	Unknown
 	qwindows.dll!00007ff81de729df()	Unknown
 	Qt5Core.dll!000000005e2b6a65()	Unknown
 	[External Code]	
 	Qt5Core.dll!000000005e2b61d9()	Unknown
 	qwindows.dll!00007ff81de729b9()	Unknown
 	Qt5Core.dll!000000005e26bf91()	Unknown
 	Qt5Core.dll!000000005e26e477()	Unknown
 	interface.exe!main(int argc, const char * * argv) Line 250	C++
 	[External Code]

**Comment by huffman** [comment]
Thanks @Nex-Pro!

It looks like we already have a bug opened for that - actually, it looks like you hit this while testing another PR too. I added this case to the bug report.

**Comment by huffman** [comment]
@Nex-Pro I also added your suggestions for improvements to FogBugz
